// expo-iap/build/types.js
// ============================================================================
// AUTO-GENERATED TYPES — DO NOT EDIT DIRECTLY
// Run `bun run generate:types` after updating any *.graphql schema file.
// ============================================================================
export var ErrorCode;
(function (ErrorCode) {
    ErrorCode["ActivityUnavailable"] = "activity-unavailable";
    ErrorCode["AlreadyOwned"] = "already-owned";
    ErrorCode["AlreadyPrepared"] = "already-prepared";
    ErrorCode["BillingResponseJsonParseError"] = "billing-response-json-parse-error";
    ErrorCode["BillingUnavailable"] = "billing-unavailable";
    ErrorCode["ConnectionClosed"] = "connection-closed";
    ErrorCode["DeferredPayment"] = "deferred-payment";
    ErrorCode["DeveloperError"] = "developer-error";
    ErrorCode["EmptySkuList"] = "empty-sku-list";
    ErrorCode["FeatureNotSupported"] = "feature-not-supported";
    ErrorCode["IapNotAvailable"] = "iap-not-available";
    ErrorCode["InitConnection"] = "init-connection";
    ErrorCode["Interrupted"] = "interrupted";
    ErrorCode["ItemNotOwned"] = "item-not-owned";
    ErrorCode["ItemUnavailable"] = "item-unavailable";
    ErrorCode["NetworkError"] = "network-error";
    ErrorCode["NotEnded"] = "not-ended";
    ErrorCode["NotPrepared"] = "not-prepared";
    ErrorCode["Pending"] = "pending";
    ErrorCode["PurchaseError"] = "purchase-error";
    ErrorCode["PurchaseVerificationFailed"] = "purchase-verification-failed";
    ErrorCode["PurchaseVerificationFinishFailed"] = "purchase-verification-finish-failed";
    ErrorCode["PurchaseVerificationFinished"] = "purchase-verification-finished";
    ErrorCode["QueryProduct"] = "query-product";
    ErrorCode["ReceiptFailed"] = "receipt-failed";
    ErrorCode["ReceiptFinished"] = "receipt-finished";
    ErrorCode["ReceiptFinishedFailed"] = "receipt-finished-failed";
    ErrorCode["RemoteError"] = "remote-error";
    ErrorCode["ServiceDisconnected"] = "service-disconnected";
    ErrorCode["ServiceError"] = "service-error";
    ErrorCode["SkuNotFound"] = "sku-not-found";
    ErrorCode["SkuOfferMismatch"] = "sku-offer-mismatch";
    ErrorCode["SyncError"] = "sync-error";
    ErrorCode["TransactionValidationFailed"] = "transaction-validation-failed";
    ErrorCode["Unknown"] = "unknown";
    ErrorCode["UserCancelled"] = "user-cancelled";
    ErrorCode["UserError"] = "user-error";
})(ErrorCode || (ErrorCode = {}));
// -- End subscription helper types
